### ISSUE TITLE
solid:oidcIssuer discovery

### ISSUE THREAD

**Issue by elf-pavlik** [solid/solid-auth-client#107]
https://github.com/solid/webid-oidc-spec#issuer-discovery-from-webid-profile

> During the Provider Selection or Provider Confirmation steps it is necessary to discover, for a given WebID, the URI of the authorized OIDC provider for that WebID.

I have currently initial deployment of NSS5.0.0-beta.6 on one domain to act as Resource Server holding my WebID Profile. I also have another deployment of NSS5.0.0-beta.6 to act as Identity Provider (OIDC Issuer). When I test latest master of this repo and enter my WebID in the popup, it makes request directly to `/authorize` of server hosting my WebID and doesn't try to discover `solid:oidcIssuer` from my WebID profile.

**Comment by dmitrizagidulin** [comment]
Ah, interesting. I think I know why that is (has to do with design assumptions). I’ll take a look at the code, to be sure.

**Comment by elf-pavlik** [comment]
It looks that indeed [`webid-oidc.js'](https://github.com/solid/solid-auth-client/blob/f9ac64cf0b6408ee293ba4943a6b4dd4a5960eb9/src/webid-oidc.js) makes some assumptions it shouldn't.

* [`login()'](https://github.com/solid/solid-auth-client/blob/f9ac64cf0b6408ee293ba4943a6b4dd4a5960eb9/src/webid-oidc.js#L18)
  * [`getRegisteredRp()`](https://github.com/solid/solid-auth-client/blob/f9ac64cf0b6408ee293ba4943a6b4dd4a5960eb9/src/webid-oidc.js#L99)
    * [`registerRp()`](https://github.com/solid/solid-auth-client/blob/f9ac64cf0b6408ee293ba4943a6b4dd4a5960eb9/src/webid-oidc.js#L133)

```js
const registration = {
    issuer: idp,
    grant_types: ['implicit'],
    redirect_uris: [callbackUri],
    response_types: [responseType],
    scope: 'openid profile'
}
```

While [`IdpSelect.js`](https://github.com/solid/solid-auth-client/blob/f9ac64cf0b6408ee293ba4943a6b4dd4a5960eb9/popup-app/components/IdpSelect.js) has
```html
<p>Please enter your WebID or the URL of your identity provider:</p>
```
And [`handleSelectIdp()`](https://github.com/solid/solid-auth-client/blob/f9ac64cf0b6408ee293ba4943a6b4dd4a5960eb9/popup-app/components/IdpSelect.js#L43) just passes the WebID calling
```js
await auth.login(idp, loginOptions)
```

What do you think if we had `login(webIdOrIdp)` and first step would check if de-referenced returns WebID Profile and in that case do discovery based on `solid:oidcIssuer` ?

**Comment by elf-pavlik** [comment]
I started playing with simple implementation, which adds a function to `webid-oidc.js` to discover OIDC Issuer from WebID Profile and gets used as first step of `WebIdOidc.login()`. It needs to parse turtle, currently I use https://github.com/rdfjs/parser-n3 but preferably it should use dependency injection so that application can provide a parser which it already imports. Do you think it should get injected using `SolidAuthClient` constructor?

**Comment by RubenVerborgh** [comment]
We have such a function somewhere, possibly solid auth manager.

**Comment by elf-pavlik** [comment]
I guess this one: https://github.com/solid/oidc-auth-manager/blob/master/src/preferred-provider.js#L22
It also seems to make incorrect assumptions. WebID Profile can stay hosted on server also hosting OIDC provider but in the WebID Profile one can specify other issuer. I think it has to check the WebID Profile first and if it doesn't include `solid:oidcIssuer` statement then fall back to `/.well-known` based discovery.

I also see it imports `rdflib` while I only needed https://github.com/rdfjs/parser-n3 and checked for statement with `solid:oidcIssuer` directly on `data` event during parsing.

**Comment by RubenVerborgh** [comment]
True, I think we can do better indeed 🙂 Constructor param maybe a little
too far for now; overhead should be minimal.

**Comment by dmitrizagidulin** [comment]
> It also seems to make incorrect assumptions. WebID Profile can stay hosted on server also hosting OIDC provider but in the WebID Profile one can specify other issuer. I think it has to check the WebID Profile first 

So just for background - that assumption was basically a performance optimization. The WebID profile only gets checked for an external oidcIssuer only if it's different from the `serverUri`, not on every request, since I think that might be excessive.

**Comment by elf-pavlik** [comment]
I don't see a clean way of injecting turtle parser dependency. This module exports *a singleton instance of SolidAuthClient*. If it exported the class I could imagine application importing it to simply: 
```js
const auth = new SolidAuthClient({ turtleParser: new N3Parser() })
```
and later in `SolidAuthClient#login` it could pass it to `WebIdOidc#login`

**Comment by elf-pavlik** [comment]
If this module exports the class, binding methods to instance could happen in the constructor, possibly using https://github.com/sindresorhus/auto-bind

**Comment by RubenVerborgh** [comment]
webpack can just do it, no?

**Comment by elf-pavlik** [comment]
I don't understand how webpack can provide a clear solution. I would like to develop Solid based Progressive Web App. In that application I would like to import:
* `solid-auth-client`
* `@rdfjs/parser-n3`
* `quadstore`
* etc.

If the application has useful features without login, I will most likely use dynamic `import()` to lazy load `solid-auth-client` once user decides to login. On the other hand `@rdfjs/parser-n3` will get loaded much earlier in the application shell. Once I `import()` `solid-auth-client` I imagine creating an instance of it this way:

```js
import ParserN3 from '/node_modules/rdfjs/parser-n3/index.js'
// ...

// when user decides to login
const SolidAuthClient = await import('/node_modules/solid-auth-client/module/index.js')
const auth = new SolidAuthClient({ turtleParser: new ParserN3() })
```

This would work in a similar way as modules which depend on LevelDOWN, for example: https://www.npmjs.com/package/quadstore#rdfstore-class

```js
import memdown from 'memdown'
import { RdfStore } from 'quadstore'
const store = new RdfStore(memdown())
```

or using different LevelDOWN implementation

```js
import leveldown from 'leveldown'
import { RdfStore } from 'quadstore'
const store = new RdfStore(leveldown('.db'))
```

**Comment by RubenVerborgh** [comment]
> I don't understand how webpack can provide a clear solution.

So webpack could either a) bundle a different dependency b) leave the dependency open, for the scenario you describe above.

But indeed, your scenario with multi-stage loading (which I like BTW) is more complex. I like the idea of the `window.solid` namespace, which currently has `solid.auth` and `solid.data`. If something does not ship with an auth implementation, it just uses `solid.auth` (thanks to webpack); we could do the same with `solid.rdfjs` for instance.
